### PR TITLE
[OpenTelemetry] Fixed dropped self-diagnostics

### PR DESCRIPTION
**Pull request by martincostello**
## Changes

Fix self-diagnostic logs being dropped if they contained too many 3-byte characters. Now the content is only truncated.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] ~~Changes in public API reviewed (if applicable)~~

### CODE REVIEW

File: src/OpenTelemetry/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,11 @@ Notes](../../RELEASENOTES.md).
 
 ## Unreleased
 
+* Fixed self-diagnostics log lines being silently dropped when an event message
+  or parameter contained enough 3-byte UTF-8 characters to overflow the internal
+  buffer estimate. Such content is now truncated.
+  ([#TODO](https://github.com/open-telemetry/opentelemetry-dotnet/pull/TODO))
```
Comment (martincostello) [inline src/OpenTelemetry/CHANGELOG.md]:
```suggestion
  ([#7543](https://github.com/open-telemetry/opentelemetry-dotnet/pull/7543))
```

**Review by stevejgordon — APPROVED** [review]
Looks good as proposed

File: test/OpenTelemetry.Tests/Internal/SelfDiagnosticsEventListenerTests.cs [hunk 2]
```diff
@@ -266,6 +268,67 @@ public void SelfDiagnosticsEventListener_EncodeInBuffer_IsParameter_NotEvenSpace
...
+    [Fact]
+    public void SelfDiagnosticsEventListener_EncodeInBuffer_ThreeByteUtf8_TruncatesWithoutThrowing()
+    {
+        var buffer = new byte[20];
+        var str = new string(ThreeByteChar, 8); // 8 * 3 = 24 bytes, does not fit in 20.
+
+        var endPos = SelfDiagnosticsEventListener.EncodeInBuffer(str, false, buffer, 0);
+
+        var written = Encoding.UTF8.GetString(buffer, 0, endPos);
+        Assert.StartsWith(ThreeByteChar.ToString(), written, StringComparison.Ordinal);
```
Comment (xiang17) [inline test/OpenTelemetry.Tests/Internal/SelfDiagnosticsEventListenerTests.cs:280]: nit: would it be better to verify the exact number of `ThreeByteChar`s?

**Review by xiang17 — APPROVED** [review]
Looks good. This rare corner case has been causing silent drops that's hard to find.

**Review by rajkumar-rangaraj — APPROVED** [review]
